hera_qm/firstcal_metrics.py

```diff
@@ -0,0 +1,544 @@
+        filename : str
+            filename to read in
+
+        """
+        # get filetype
+        filetype = filename.split('.')[-1]
+        if filetype == 'json':
+            with open(filename, 'r') as f:
+                self.metrics = json.load(f)
+
+            # ensure keys of dicts are not strings
+            for i in self.metrics['ant_avg'].keys():
+                self.metrics['ant_avg'][int(i)] = self.metrics['ant_avg'].pop(i)
+            for i in self.metrics['ant_std'].keys():
+                self.metrics['ant_std'][int(i)] = self.metrics['ant_std'].pop(i)
+            for i in self.metrics['time_std'].keys():
+                self.metrics['time_std'][float(i)] = self.metrics['time_std'].pop(i)
+            for i in self.metrics['z_scores'].keys():
+                self.metrics['z_scores'][float(i)] = self.metrics['z_scores'].pop(i)
+
+        elif filetype == 'pkl':
+            with open(filename, 'rb') as f:
+                inp = pkl.Unpickler(f)
+                self.metrics = inp.load()
+        else:
+            raise IOError("Filetype not recognized, try a json or pkl file")
+
+    def delay_std(self, return_dict=False):
+        """
+        Calculate standard deviations of per-antenna delay solutions
+        and aggregate delay solutions. Assign a z-score for
+        each (antenna, time) delay solution w.r.t. aggregate
+        mean and standard deviation.
+
+        Uses sqrt( astropy.stats.biweight_midvariance ) for a robust measure
+        of the std
+
+        Input:
+        ------
+
+        return_dict : bool, [default=False]
+            if True, return time_std, ant_std and z_scores
+            as a dictionary with "ant" as keys
+            and "times" as keys
+            else, return as ndarrays
+
+        Output:
+        --------
+        return ant_avg, ant_std, time_std, agg_std, z_scores
+
+        ant_avg : ndarray, shape=(N_ants,)
+            average delay solution across time for each antenna
+
+        ant_std : ndarray, shape=(N_ants,)
+            standard deviation of delay solutions across time
+            for each antenna
+
+        time_std : ndarray, shape=(N_times,)
+            standard deviation of delay solutions across ants
+            for each time-stamp
+
+        agg_std : float
+            aggregate standard deviation of delay solutions
+            across all antennas
+
+        z_scores : ndarray, shape=(N_ant, N_times)
+            absolute value of z_scores (standard scores)
+            for each (antenna, time) delay solution w.r.t. agg_std
+
+        """
+        # calculate standard deviations
+        ant_avg = self.delay_avgs
+        ant_std = np.sqrt( astats.biweight_midvariance(self.delay_offsets, axis=1) )
+        time_std = np.sqrt( astats.biweight_midvariance(self.delay_offsets, axis=0) )
+        agg_std = np.sqrt( astats.biweight_midvariance(self.delay_offsets) )
+
+        # calculate z-scores
+        z_scores = np.abs(self.delay_offsets / agg_std)
```

Comment:
Why are we taking the absolute value? It might be better to leave the sign information in, and then take the absolute value when you go to plot or make a cut. But in principle I think this throws away information that could be useful.